### ISSUE TITLE
"csc.exe" exited with code -2146232797

### ISSUE THREAD

**Issue by andrey-messi** [dotnet/roslyn#23290]
**Version Used**: vs2017 version 15.4.4

**Case 1**

**Steps to Reproduce**:

1. Try compile this code:
```using System;
namespace Ns
{
    internal static class Program
    {
        private static void Main()
        {
            Validate(1);
        }
        static int Validate(int a)
        {
            int Func1()
            {
                int Func2()
                {
                    int Func3(int y)
                    {
                        return Func2();
                    }
                    return Func3(a);
                }
                return Func2();
            }
            return Func1();
        }
    }
}
```

**Expected Behavior**: Successfully compiled

**Actual Behavior**: COMPILE ERROR -- "csc.exe" exited with code -2146232797

**Case 2**

**Steps to Reproduce**:

1. Get code from case 1 and change a call to Func3(a) with a call to Func3(1) 
Code with this change is shown below, try compile this code:
`using System;
namespace Ns
{
    internal static class Program
    {
        private static void Main()
        {
            Validate(1);
        }
        static int Validate(int a)
        {
            int Func1()
            {
                int Func2()
                {
                    int Func3(int y)
                    {
                        return Func2();
                    }
                    return Func3(1);
                }
                return Func2();
            }
            return Func1();
        }
    }
}
`

**Expected Behavior**: Successfully compiled

**Actual Behavior**: Successfully compiled (that means that a bug in a case 1 is in somewhere where a closure to variable 'a' is captured)

**Case 3**

**Steps to Reproduce**:

1. Get code from case 1 and change a call to Func2() (within a Func2() function) with a call to Func1() 
Code with this change is shown below, try compile this code:
`using System;
namespace Ns
{
    internal static class Program
    {
        private static void Main()
        {
            Validate(1);
        }
        static int Validate(int a)
        {
            int Func1()
            {
                int Func2()
                {
                    int Func3(int y)
                    {
                        return Func1();
                    }
                    return Func3(a);
                }
                return Func2();
            }
            return Func1();
        }
    }
}
`

**Expected Behavior**: Successfully compiled

**Actual Behavior**: Successfully compiled as well as in a case 2. ( even though code still has a closure to variable 'a' captured)

**Comment by MaStr11** [comment]
A test with [sharplab.io](http://sharplab.io) reveals that this compiles with the current master branch but fails with the *default* compiler.

**Comment by jcouv** [comment]
Assigning to @agocke to confirm this was indeed fixed by the 15.5 redesign of local function internals.

**Comment by agocke** [comment]
Yup, this appears to be fixed.
